Hyperion/SerializerFactories/EnumerableSerializerFactory.cs

```diff
@@ -68,33 +75,84 @@ private static Type GetEnumerableType(Type type)
+            var castArg = Expression.Convert(arg, argType);
+            var call = Expression.Call(castInstance, method, new Expression[] { castArg });
+            var lambda = Expression.Lambda<Action<object, object>>(call, instance, arg);
+            var compiled = lambda.Compile();
+            return compiled;
+        }
+
         public override ValueSerializer BuildSerializer(Serializer serializer, Type type,
             ConcurrentDictionary<Type, ValueSerializer> typeMapping)
         {
             var x = new ObjectSerializer(type);
             typeMapping.TryAdd(type, x);
+
             var preserveObjectReferences = serializer.Options.PreserveObjectReferences;
 
             var elementType = GetEnumerableType(type) ?? typeof(object);
             var elementSerializer = serializer.GetSerializerByType(elementType);
 
             var countProperty = type.GetTypeInfo().GetProperty("Count");
-            var addRange = type.GetTypeInfo().GetMethod("AddRange");
-            var add = type.GetTypeInfo().GetMethod("Add");
+            var constructor = GetEnumerableConstructor(type);
+            var construct = constructor != null 
+                ? CompileCtorToDelegate(constructor, elementType.MakeArrayType()) 
+                : null;
+            var addRangeMethod = type.GetTypeInfo().GetMethod("AddRange");
+            var addRange = construct == null && addRangeMethod != null
+                 ? CompileMethodToDelegate(addRangeMethod, type, elementType.MakeArrayType())
+                 : null;
+            var addMethod = type.GetTypeInfo().GetMethod("Add");
+            var add = construct == null && addRange == null && addMethod != null
+                ? CompileMethodToDelegate(addMethod, type, elementType)
+                : null;
 
             Func<object, int> countGetter = o => (int)countProperty.GetValue(o);
-
-
+            
             ObjectReader reader = (stream, session) =>
             {
-                var instance = Activator.CreateInstance(type);
-                if (preserveObjectReferences)
+                var count = stream.ReadInt32(session);
+                if (construct != null)
                 {
-                    session.TrackDeserializedObject(instance);
+                    var enumerable = Array.CreateInstance(elementType, count);
+                    for (var i = 0; i < count; i++)
+                    {
+                        var value = stream.ReadObject(session);
+                        enumerable.SetValue(value, i);
+                    }
+                    var instanceFromConstructor = construct(enumerable);
+                    if (preserveObjectReferences)
+                    {
+                        session.TrackDeserializedObject(instanceFromConstructor);
+                    }
+                    return instanceFromConstructor;
                 }
 
-                var count = stream.ReadInt32(session);
-
+                var instance = Activator.CreateInstance(type);
```

Comment:
If you're going to call `Activator.CreateInstance` make sure, that you've checked that in `CanSerialize` you've checked that this type has default constructor. Also in hyperion we accept calling non-public constructors, so `Activator.CreateInstance(type, true)` is probably better and more explicit in this context.

Reply:
Good catch on using `Activator.CreateInstance(type, true)`. On a similar note, should I provide the appropriate `BindingFlags` value to `type.GetTypeInfo().GetConstructor` to return non-public constructors there as well?